crates/proto/src/rr/dnssec/tbs.rs

```diff
@@ -7,19 +7,228 @@
 
 //! hash functions for DNSSEC operations
 
-use std::borrow::Borrow;
+use time::OffsetDateTime;
 
 use crate::{
     error::*,
-    rr::{dnssec::Algorithm, DNSClass, Name, Record, RecordType, SerialNumber},
+    rr::{dnssec::Algorithm, DNSClass, Name, Record, RecordSet, RecordType, SerialNumber},
     serialize::binary::{BinEncodable, BinEncoder, EncodeMode},
 };
 
-use super::rdata::{sig, RRSIG, SIG};
+use super::{
+    rdata::{sig, RRSIG, SIG},
+    SigSigner,
+};
 
 /// Data To Be Signed.
 pub struct TBS(Vec<u8>);
 
+impl TBS {
+    /// Returns the to-be-signed serialization of the given record set using the information
+    /// provided from the RRSIG record.
+    ///
+    /// # Arguments
+    ///
+    /// * `rrsig` - SIG or RRSIG record, which was produced from the RRSet
+    /// * `records` - RRSet records to sign with the information in the `rrsig`
+    ///
+    /// # Return
+    ///
+    /// binary hash of the RRSet with the information from the RRSIG record
+    pub fn from_rrsig<'a>(
+        rrsig: &Record<RRSIG>,
+        records: impl Iterator<Item = &'a Record>,
+    ) -> ProtoResult<Self> {
+        Self::from_sig(rrsig.name(), rrsig.dns_class(), rrsig.data(), records)
+    }
+
+    /// Returns the to-be-signed serialization of the given record set using the information
+    /// provided from the SIG record.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - labels of the record to sign
+    /// * `dns_class` - DNSClass of the RRSet, i.e. IN
+    /// * `sig` - SIG or RRSIG record, which was produced from the RRSet
+    /// * `records` - RRSet records to sign with the information in the `rrsig`
+    ///
+    /// # Return
+    ///
+    /// binary hash of the RRSet with the information from the RRSIG record
+    pub fn from_sig<'a>(
+        name: &Name,
+        dns_class: DNSClass,
+        sig: &SIG,
+        records: impl Iterator<Item = &'a Record>,
+    ) -> ProtoResult<Self> {
+        Self::new(
+            name,
+            dns_class,
+            sig.num_labels(),
+            sig.type_covered(),
+            sig.algorithm(),
+            sig.original_ttl(),
+            sig.sig_expiration(),
+            sig.sig_inception(),
+            sig.key_tag(),
+            sig.signer_name(),
+            records,
+        )
+    }
+
+    /// Returns the to-be-signed serialization of the given record set.
+    ///
+    /// # Arguments
+    ///
+    /// * `inception` - the epoch seconds of when this hashed signature will be valid
+    /// * `rr_set` - RRSet to sign
+    /// * `zone_class` - DNSClass, i.e. IN, of the records
+    /// * `signer` - the signer to use for signing the RRSet
+    ///
+    /// # Returns
+    ///
+    /// the binary hash of the specified RRSet and associated information
+    // FIXME: OMG, there are a ton of asserts in here...
```

Comment:
I think this comment applies only to the `new` constructor? But yes, there are a lot of `assert!`s for the `emit` calls.

Reply:
Cleaned it up.